ENH: Data Interface to simplify numpy type conversion

bfcondon opened the pull request:
As a follow-up to the discussion in #3027, many functions have code that converts user provided data objects into the internal data representation used by that part of the library.  This is an initial implementation of a DataInterface class to give developers control over that process.

Closes #3027 

@josef-pkt I know you are on vacation, but I would appreciate any comments when you have time.

bashtage commented on statsmodels/tools/data_interface.py:
```diff
@@ -0,0 +1,136 @@
+        else:
+            raise TypeError('Type conversion from {} to {} is not possible.'.format(internal_type, self.external_type))
+
+    def to_numpy(self, data):
```
Might be btter named as `asarray` since a recarray is a numpy type.

bfcondon replied:
what do you think of to_numpy_array so as not to confuse it with the numpy asarray function, since it does not do the same thing?

josef-pkt commented on statsmodels/base/model.py:
```diff
@@ -743,34 +743,14 @@ def predict(self, exog=None, transform=True, *args, **kwargs):
-        if exog is not None:
-            exog = np.asarray(exog)
-            if exog.ndim == 1 and (self.model.exog.ndim == 1 or
-                                   self.model.exog.shape[1] == 1):
```
I think this reshape is too specific to predict to put it into any generic function.

bashtage replied:
Interface could expose an property that is atleast2d or similar.  Alternatively it could use a keyword to always expose a minimum number of dimensions, e.g.

```
exog_interface = NumPyInterface(data=exog, model=self.model, use_formula=transform, min_dims=2)
```

josef-pkt replied:
The problem here is that we don't just want 2-dim. We want either a 2d row or a 2d column, and that depends on the model information, i.e. column vector if we only have one explanatory variable exog, otherwise row vector.)

josef-pkt commented on statsmodels/tools/data_interface.py:
```diff
@@ -0,0 +1,136 @@
+
+class DataInterface(object):
+
+    def __init__(self, permitted_types, internal_type, data=None, external_type=None, model=None, use_formula=False):
```
I think `internal_type` should default to `np.ndarray` which is currently almost always the case

josef-pkt commented on statsmodels/tools/data_interface.py:
```diff
@@ -0,0 +1,136 @@
+
+        elif from_type == pd.Series:
+
+            to_return = data.values
```
is data.values equivalent to np.asarray(data)?

bfcondon replied:
pandas returns self._data.asmatrix()

josef-pkt commented on statsmodels/tools/data_interface.py:
```diff
@@ -0,0 +1,136 @@
+        self.internal_type = internal_type
+        self.columns = getattr(data, 'columns', None)
+        self.name = getattr(data, 'name', None)
+        self.index = getattr(data, 'index', None)
```
This is a bit dangerous/fragile: You use the index from the data in the `__init__`, but the actual data in the methods could have a different `index`.

bfcondon replied:
I will remove this

bfcondon commented:
Thanks for the comments. I am on vacation for a week and will respond when I return.

bfcondon commented:
@josef-pkt @bashtage I have addressed your comments and have almost all of the tests passing.

I have a few questions:
1.  Two of the failing tests relate to #2924.  Should I include the workaround relating to hasattr(data, 'predicted_values') or wait until this is resolved?
2. Sometimes the returned result is a DesignMatrix.  To get the tests to pass I just return this unmodified.  Is this the right approach?

The two remaining failing tests relate to receiving a 1D row vector and returning a 1D column vector.  I plan to check for this and transpose the returned data if this is an issue.

bfcondon commented:
@josef-pkt If you have time, do you have any thoughts on my questions above?  I made some commits to fix it, but I am still having some transpose issues.

bfcondon commented:
@josef-pkt Other than some new scipy bugs, I have squashed the remaining issues.  The one that remains relates to bunches being returned by Phreg predict.

bfcondon commented:
Closes #1148
Closes #2133 

@josef-pkt  I added functionality to convert pandas datatypes into a matrix of dummy variables.  The test suite I copied from test_tools mostly passes except for some shape errors.  Also, for some reason there is a merge conflict.

bfcondon commented:
@bashtage @josef-pkt I fixed the failing tests for the categorical function, but there is still a merge conflict.

I know that you have been busy, but I would appreciate feedback on whether I should continue.  If you do not believe so, just let me know.
